### PR TITLE
Entailment pr

### PR DESCRIPTION
**Pull request by vyskoto4**
The entailment model implementation is not 100% final, I had to replace some hardcoded  code  by the SequenceSplit layer and the performance is not as good as before.

### CODE REVIEW

File: models/entail1509.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def entailment_embedding(model, inputs,N=608, spad=60, l2reg=1e-4):
+    model.add_node(Lambda(get_H_n, output_shape=(N,)), name='h_n', input=inputs[1])
+    model.add_node(SplitSequence(spad), name='Y', input=inputs[0])
+    model.add_node(Dense(N,W_regularizer=l2(l2reg)),name='Wh_n', input='h_n')
+    model.add_node(RepeatVector(spad), name='Wh_n_cross_e', input='Wh_n')
+    model.add_node(TimeDistributedDense(N,W_regularizer=l2(l2reg)), name='WY', input='Y')
+    model.add_node(Activation('tanh'), name='M', inputs=['Wh_n_cross_e', 'WY'], merge_mode='sum')
+    model.add_node(TimeDistributedDense(1,activation='softmax'), name='alpha', input='M')
```
Comment (pasky) [inline models/entail1509.py]: softmax on 1 dim?

File: models/entail1509.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+def generate_get_Y(size):
+    def get_Y_generator(X):
+        return get_Y(X,size)
+    return get_Y_generator
+
+def get_Y(X, size):
+    return X[:, :size, :]
+
+def get_H_0():
+    return X[:, 0, :]
```
Comment (pasky) [inline models/entail1509.py]: Unused? Anything else unused?

File: models/entail1509.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        input_shape[1]=self.split_ind
+        return input_shape
+
+
+def get_R(X):
+    Y, alpha = X.values()  
+    ans=K.T.batched_dot(Y,alpha)
+    return ans
+
+def rnn_input(model, N, spad, input,c):
```
Comment (pasky) [inline models/entail1509.py]: Can we use rnn_input from kerasts?

File: models/entail1509.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                                    return_sequences=True, go_backwards=True,
+                                    dropout_W=c['dropoutfix_inp'], dropout_U=c['dropoutfix_rec']))
+        outputs=['e0s_', 'e1s_']
+        model.add_node(name='rnndrop', inputs=['forward', 'backward'], merge_mode='concat' ,
+                              layer=Dropout(c['dropout'], input_shape=(2*spad, int(N*c['sdim'])) ))
+        return ['rnndrop']*2
+
+
+def entailment_embedding(model, inputs,N=608, spad=60, l2reg=1e-4):
+    model.add_node(Lambda(get_H_n, output_shape=(N,)), name='h_n', input=inputs[1])
```
Comment (pasky) [inline models/entail1509.py]: Can we rename get_H_n?

File: models/entail1509.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    model.add_node(SplitSequence(spad), name='Y', input=inputs[0])
```
Comment (pasky) [inline models/entail1509.py]: Maybe FirstSequence is better name?

File: models/entail1509.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    model.add_node(Dense(N,W_regularizer=l2(l2reg)), name='Wh', input='h_n')
+    outputs=['Wr','Wh']
+    return outputs
+
+
+def prep_model(model, N, s0pad, s1pad, c):
+    model.add_node(name="embmerge", inputs=['e0','e1'], merge_mode='concat', layer=Activation('linear'))
+    rnn_outputs=rnn_input(model,N,s0pad,'embmerge',c)
+    outputs = entailment_embedding(model, rnn_outputs,2*N,s0pad,c['l2reg'])
+    final_output = B.to_n_ptscorer(model, outputs, c['Ddim'], N, c['l2reg'], pfx="entail_out", output_dim=3)
```
Comment (pasky) [inline models/entail1509.py]: Maybe we should just return outputs and leave the scoring issue task-specific.

Comment (pasky) [inline models/entail1509.py]: Also careful that we don't do two softmax in succession.

File: models/entail1509.py (hunk 1 above)
Comment (vyskoto4) [reply models/entail1509.py]: Ok, it was a typo, replaced it with linear

File: models/entail1509.py (hunk 2 above)
Comment (vyskoto4) [reply models/entail1509.py]: Unused functions should be gone now.

File: models/entail1509.py (hunk 3 above)
Comment (vyskoto4) [reply models/entail1509.py]: Yes, after some changes I made.

File: models/entail1509.py (hunk 4 above)
Comment (vyskoto4) [reply models/entail1509.py]: Ok.

File: models/entail1509.py (hunk 5 above)
Comment (vyskoto4) [reply models/entail1509.py]: I changed it to FirstSentence.

File: models/entail1509.py (hunk 6 above)
Comment (vyskoto4) [reply models/entail1509.py]: I removed the scorer and two softmax in succession.

**Comment by pasky** [comment]
Thanks a lot for the update! I'm a lot happier I think, a bit wary about the ugly looking code in rnn_input but I understand the neccessity and we'll be able to throw away that complexity again with the Keras1.0 port. So it's okay. :)

But I'm still a bit leery about the ptscorer business. I think the basic sanity check is:
- What happens when we use a different model with the rte task? (E.g. re the sigmoid -> linear activation change. But I don't understand that anyway?)
- What happens when we use the model with a different task?
  Are these cases both okay?

The other problem I see is that we use _completely_ different ptscoring strategies when the to_n is involved - I don't understand that.  I think we discussed just making the output dimension a parameter of mlp_ptscorer?  Would that be inferior to the current solution?

**Comment by vyskoto4** [comment]
- The sigmoid-linear change in RTE task has been done only because of the new model, as the reached accuracy was the best with these settings. I also tried RNN, AVG (and maybe CNN) and their performance seemed to be the same as before.
- I did not test it with a different task. That is on my todo list.
- I'l test the modified mlp_scorer as soon as I get some GPU time.

EDIT:
- I deleted the custom scorer and the mlp_scorer is being used. I did not notice any significant  performance changes after this.
- Now testing the new (ok, its actually old today) on anssel/wang. I did not run full eval but the best val MRR during traning was 0.845.
